slurmctl: add support for TLS certificates

heitorPB opened the pull request:
**Please provide description of the purpose of this pull request, as well as its
motivation and context.**

Add three configurations in slurmctld charm to allow the operator to
supply TLS certificates for etcd.

Documentation in: https://github.com/omnivector-solutions/osd-documentation/pull/72

**How was the code tested?**

New tests for TLS certs added.



Final checklist
- [x] I am the author of these changes, or I have the rights to submit them.
- [x] I added the relevant changed to the README and/or documentation.
- [x] I self reviewed my own code.
- [x] I updated the `CHANGELOG` according to the [contributing
  guide](https://omnivector-solutions.github.io/osd-documentation/master/contributing.html#changelog)

heitorPB committed:
slurmctld: add support for TLS certificates

Add three configurations in slurmctld charm to allow the operator to
supply TLS certificates for etcd. Two options for the TLS cert (pub and
private files), and one optional as the CA public certificate, in case
the nodes do not have that or it was used an internal CA. etcd will not
use that cert to verify the clients.

Add integration tests to validade the certs are in place and can be
used correctly.

matheushent commented on charm-slurmctld/src/charm.py:
```diff
@@ -495,6 +497,16 @@ def _on_write_slurm_config(self, event):
             event.defer()
             return
 
+        # check if both certificates are supplied
+        tls_key = self.model.config['tls-key']
+        tls_cert = self.model.config['tls-cert']
+        self._stored.use_tls = (bool(tls_key) and bool(tls_cert))
+        self._stored.use_tls_ca = bool(self.model.config['tls-ca-cert'])
+
+        # TODO this will fire everytime a the configuration changed, we don't
+        #      need that. This should happen only if the tls configs changed
+        self._etcd.setup_tls()
```
to implement the TODO, thw following could be done:

```suggestion
        # check if both certificates are supplied and setup TLS in case they are
        tls_key = self.model.config['tls-key']
        tls_cert = self.model.config['tls-cert']
        if not self._stored.use_tld and not self._stored.use_tls_ca:
            self._stored.use_tls = (bool(tls_key) and bool(tls_cert))
            self._stored.use_tls_ca = bool(self.model.config['tls-ca-cert'])
            if self._stored.use_tld and self._stored.use_tls_ca:
                self._etcd.setup_tls()
```

A bit strange? yes, but I guess the logic is it

heitorPB replied:
We have 3 different states:
- no TLS
- only TLS certs
- TLS certs + CA cert

We only need to run the function `setup_tls` when changing from one state to another, we have 6 different transitions, so we need a logic a bit more complex than this one.

heitorPB replied:
This is not critical at all, the `setup_tls` function has a safeguard. This task took longer than I anticipated, I don't think it is worth implementing this logic right now.

heitorPB replied:
This is not a problem at all, `setup_tls()` will not cause any harm. I don't think we should worry about this right now.

heitorPB committed:
slurcmtld: pin influxdb library to 5.3.1

heitorPB committed:
CI: fix publishing to candidate

The release variable must not be escaped, so removing the double quotes.
Otherwise, charmcraft receives "edge --release=candidate" as the release
name, which is incorrect.

julianaklulo reviewed (approved):
Nice work!

heitorPB commented:
Rebased to squash the commits.
